man/scr_halt.1
.TH scr_halt 1  "" "SCR" "SCR"

.SH NAME
scr_halt \- set or modify halt conditions of an SCR job

.SH SYNOPSIS
.B scr_halt [options] [jobid ...]

.SH DESCRIPTION
The \fBscr_halt\fR command is used to instruct an actively running SCR
job to safely exit.  The job can be instructed to exit after its next
successful checkpoint, exit before or after a specified time, or exit
immediately.

.SH OPTIONS
.TP
.BI "-z, --all"
Halt all jobs on the system.
.TP
.BI "-u, --user " LIST
Halt jobs for a comma-separated \fILIST\fR of users.
.TP
.BI "-c, --checkpoints " N
Halt job after \fIN\fR checkpoints.
.TP
.BI "-b, --before " TIME
Halt job before specified \fITIME\fR.  Uses SCR_HALT_SECONDS if set.
.TP
.BI "-a, --after " TIME
Halt job after specified \fITIME\fR.
.TP
.BI "-i, --immediate"
Halt job immediately.
.TP
.BI "-s, --seconds " N
Set or reset SCR_HALT_SECONDS for active job.
.TP
.BI "-l, --list"
List the current halt conditions specified for a job or jobs.
.TP
.BI "--unset-checkpoints"
Unset any checkpoint halt condition.
.TP
.BI "--unset-before"
Unset any halt before condition.
.TP
.BI "--unset-after"
Unset halt after condition.
.TP
.BI "--unset-seconds"
Unset halt seconds.
.TP
.BI "--unset-reason"
Unset the current halt reason
.TP
.BI "-r, --remove"
Remove halt file from nodes in job allocation.
.TP
.BI "-e, --every"
Write halt file to every node in job allocation.
.TP
.BI "-v, --verbose"
Increase verbosity.
.TP
.BI "-h, --help"
Print usage.
.LP
LIST refers to a comma-separated list of words. All options which take a
LIST argument may also be specified multiple times
(e.g. --user=sally,tim --user=frank).
.LP
TIME arguments are parsed using the perl Date::Manip(3pm) package,
and thus may be specified in one of many formats.
Examples include '12pm', 'yesterday,noon', '12/25-15:30:33', and so on.
See the Date::Manip(3pm) manpage for more examples.
.LP
If no jobid is specified, and \fB--user\fR and \fB--all\fR are not given,
then SLURM_JOBID is used if set.  This feature is convenient when invoking
the command in job batch scripts.

.SH EXAMPLES
.TP
(1) Instruct job 1234 to exit after completing its next successful checkpoint:
.nf
>> scr_halt 1234
.fi
.TP
(2) Instruct all my jobs (username3) to exit after completing their next successful checkpoint:
.nf
>> scr_halt -u username3
.fi
.TP
(3) Instruct job 1234 to exit after completing 10 checkpoints:
.nf
>> scr_halt -c 10 1234
.fi
.TP
(4) Instruct job 1234 to exit after "12:00pm today":
.nf
>> scr_halt -a '12:00pm today' 1234
.fi
.TP
(5) Instruct job 1234 to exit before "8:30am tomorrow":
.nf
>> scr_halt -b '8:30am tomorrow' 1234
.fi
.TP
(6) Instruct job 1234 to exit before "12pm Aug 10":
.nf
>> scr_halt -b '12pm 8/10' 1234
.fi
.TP
(7) Instruct job 1234 to exit immediately:
.nf
>> scr_halt -i 1234
.fi
.TP
(8) Instruct job 1234 to exit after 10 checkpoints or before "8:30am tomorrow", which ever comes earlier:
.nf
>> scr_halt -c 10 -b '8:30am tomorrow' 1234
.fi
.TP
(9) Set SCR_HALT_SECONDS to 1800 on job 1234:
.nf
>> scr_halt -s 1800 1234
.fi
.TP
(10) Remove the halt file from job 1234:
.nf
>> scr_halt -r 1234
.fi

.SH SEE ALSO
\fISCR User Manual\fR
.LP
\fBscr\fR(1), \fBscr_srun\fR(1), \fBscr_prerun\fR(1),
\fBscr_postrun\fR(1), \fBscr_index\fR(1)


ChangeLog
2012-02-29 Adam Moody <moody20@llnl.gov>

	* scr_util_mpi.c : Added some functions to send/recv strings via MPI.

	* scr_reddesc.c : Step through redundancy descriptor hashes by name
	as determined by rank 0 (no longer need to number CKPTDESC in
	sequential order).

2012-02-08 Adam Moody <moody20@llnl.gov>

	* m4/x_ac_expand_install_dirs.m4 : Fix to address datarootdir variable
	added from autoconf 2.59 to 2.63.

2012-02-07 Adam Moody <moody20@llnl.gov>

	* scr_param.c : Added hash to return pointers to values returned from
	getenv, return those values directly caused seg faults on cray
	systems.

	* scr_globals.c : Removed scr_hop_distance, added, scr_group, scr_comm_node,
	and scr_comm_node_across.

2012-02-01 Adam Moody <moody20@llnl.gov>

	* scr_util.c : Added scr_strdupf.

2012-01-31 Adam Moody <moody20@llnl.gov>

	* : split scr.c into component files and moved global variables to scr_globals.c

	* : Dropped scr_comm_local/level and associated parameters and replaced
	  them with group descriptors (defines high-level groups of processes)
	  and store descriptors (defines parameters about storage locations).

2011-08-31 Adam Moody <moody20@llnl.gov>

	* scr_hash_util.h/c : Added scr_hash_set/get_double.

	* scr_transfer.c : Only bother writing to the transfer file if we
	  successfully opened it.

	* scr.c : Added code to support async flush in 1.1-8.

2011-08-03 Adam Moody <moody20@llnl.gov>

	* : Recording meta data in filemap and summary files instead of separate
	  meta data files.

	* scr_filemap.h,
	  scr_filemap.c : Added routines to set/get/unset meta for each file.

2011-07-30 Adam Moody <moody20@llnl.gov>

	* : Changed filemap, summary, and index files to refer to dataset ids
	  rather than checkpoint ids.  Rebuild now rebuilds all datasets in cache
	  instead of just the most recent.  Not yet fully tested.

	* : Moved halt, flush, and nodes files from control directory to prefix
	  directory.  Now only written by rank 0 rather than all master nodes.

	* scr_scavenge.in : Renamed scr_flush to scr_scavenge and changed verbage
	  from flush and drain to scavenge where appropriate.  Flush was overloaded
	  to refer to different concepts and drain soon will be.

	* scr_hash.h,
	  scr_hash.c : Moved function to list integer keys from filemap to hash.
	  Fixed bug which was sorting integer keys as strings in scr_hash_sort_int().

2011-06-28 Adam Moody <moody20@llnl.gov>

	* scr_env.h,
	  scr_env.c : Added file to isolate typical machine dependent code.

	* scr_dataset.h,
	  scr_dataset.c : Associates attributes with a set of files.  Extends
	  SCR to handle general output datasets rather than just checkpoints.

	* scr_err.h,
	  scr_err.c,
	  scr_err_serial.c : Write all messages to stdout (errors were going to stderr).
	  Include version number in all messages.  Added new scr_warn() function.

	* scr_io.h,
	  scr_io.c : Added functions to return current working directory and
	  to build the absolute path to a file.  Also added functions to compress
	  and decompress files.

	* scr_util.h,
	  scr_util.c : Added functions to pack and unpack integer types into a
	  buffer in network order.

	* scr_hash_util.h,
	  scr_hash_util.c : Added functions to set/get integer values.

	* scr_flush_file.c : Fixed bug that was returning with wrong code if
	  named checkpoint is invalid.

	* scr_hash.c : Modified code to use new pack / unpack functions.

	* scr_meta.h,
	  scr_meta.c : Modified code to use new get / set hash functions.  Also added
	  functions to record original directory specified by user at SCR_Route_file time.

	* scr_copy.c : Allow caller to specify parameters using options instead of arg order.

	* scr_dataset.h,
	  scr_dataset.c : Added functions to set / get / unset a dataset hash.

	* scr_conf.h,
	  scr.c : Added partial support for containers (available in flush and fetch,
	  but still need drain support).  Added code to record dataset attributes.  Added
	  code to remember original directories specified by caller, need support to
	  create them and delete them with checkpoint.

2011-06-10 Adam Moody <moody20@llnl.gov>

	* scr_err.h,
	  scr_err_serial.c,
	  scr.c : Added version string to error messages.

2011-05-23 Adam Moody <moody20@llnl.gov>

	* scr_hash.c,
	  tv_data_display.c,
	  tv_data_display.h : Updated to latest C++View files to restore printing capability
	  in TotalView.

2011-04-13 Adam Moody <moody20@llnl.gov>

	* scr.c : Added scr_summary_read_v4_to_v5 to read old scr_summary.txt files and
	  convert them into equivalent v5 hash for backwards compatibility.  This enables
	  v1.1-8 to restart from checkpoints stored in v1.1-7 or earlier.

	* scr_hash.c,
	  scr_hash.h : Added scr_hash_sort to sort keys in hash by string, also modified
	  integer sort call to accept ascending or descending flag.

	* scr_index.c : Added --remove option to drop a directory from the index.

	* man/scr_index.1.in : Described format of --list output and added an example.

2011-04-06 Adam Moody <moody20@llnl.gov>

	* man/scr_index.1.in : Added man page for scr_index, and removed references to
	  scr_check_complete.

	* doc/scr_users_manual.pdf,
	  doc/scr_developers_manual.pdf : Updated user's manual and renamed it.  Added
	  developer's manual.

2011-03-24 Adam Moody <moody20@llnl.gov>

	* scr.c : Fixed bug regarding leaking group handles.

	* scr.c,
	  scr_hash_mpi.c : Moved MPI related functions for hash from scr.c to scr_hash_mpi.c.
	  Includes support for a sparse global exchange.

	* m4/x_ac_gcs.m4,
	  configure.ac : Added support for libgcs (generalized comm split library), which
	  enables one to split processes on arbitrary data.  Since this library is only at
	  LLNL currently, also added --with-gcs configure option and HAVE_LIBGCS compile flag
	  to enable it.

	* scr_hash.h,
	  scr_hash.c : Added scr_hash_getf(), similar to scr_hash_setf().
	  Added hash keys for XOR file header.

2010-12-20 Adam Moody <moody20@llnl.gov>

	* scr_io.h,

scripts/pyfe/pyfe/scr_halt.py
#! /usr/bin/env python3

# scr_halt.py

import os, sys

if 'pyfe' not in sys.path:
  sys.path.insert(0, '/'.join(os.path.realpath(__file__).split('/')[:-2]))
  import pyfe

import argparse
import stat
from pyfe import scr_const
from pyfe.scr_common import runproc
from pyfe.parsetime import parsetime


def scr_halt(bindir=None,
             checkpoints=None,
             before=None,
             after=None,
             immediate=False,
             seconds=None,
             dolist=False,
             unset_checkpoints=False,
             unset_before=False,
             unset_after=False,
             unset_seconds=False,
             unset_reason=False,
             remove=False,
             verbose=False,
             dirs=[]):
  ret = 0

  if bindir is None:
    bindir = scr_const.X_BINDIR

  # path to scr_halt_cntl command
  scr_halt_cntl = os.path.join(bindir, 'scr_halt_cntl')

  # use current working directory if none specified
  if not dirs:
    dirs = [os.getcwd()]

  # list to accumulate options for scr_halt_cntl command
  halt_conditions = []

  # the -r option overrides everything else
  if not remove:
    # halt after X checkpoints
    # TODO: check that a valid value was given
    if checkpoints is not None:
      checkpoints = str(checkpoints)
      halt_conditions = ['-c', checkpoints]

    # halt before time
    if before is not None:
      secs = parsetime(before)
      halt_conditions.append('-b')
      halt_conditions.append(str(secs))

    # halt after time
    if after is not None:
      secs = parsetime(after)
      halt_conditions.append('-a')
      halt_conditions.append(str(secs))

    # set (reset) SCR_HALT_SECONDS value
    # TODO: check that a valid value was given
    if seconds is not None:
      halt_conditions.append('-s')
      halt_conditions.append(seconds)

    # list halt options
    if dolist:
      halt_conditions.append('-l')

    # push options to unset any values
    if unset_checkpoints:
      halt_conditions.append('-xc')
    if unset_before:
      halt_conditions.append('-xb')
    if unset_after:
      halt_conditions.append('-xa')
    if unset_seconds:
      halt_conditions.append('-xs')
    if unset_reason:
      halt_conditions.append('-xr')

    # if we were not given any conditions, set the exit reason to JOB_HALTED
    if len(halt_conditions) == 0 or immediate:
      halt_conditions.append('-r')
      halt_conditions.append('JOB_HALTED')

  # create a halt file in each target prefix directory
  for d in dirs:
    print('Updating halt file in ' + d)
    rc = 0

    # build the name of the halt file
    halt_file = os.path.join(d, '.scr', 'halt.scr')

    # remove the halt file and move on to next direcotry
    # if no conditions are specified
    if halt_conditions == []:
      try:
        os.remove(halt_file)
      except:
        pass
      continue

    # create scr prefix directory
    os.makedirs(os.path.join(d, '.scr'), exist_ok=True)

    # execute the command
    # create the halt file with specified conditions
    # TODO: Set halt file permissions so system admins can modify them
    ####### ensure we have chmod664 ? then also need to ensure owned path is 775 ?
    ### If a chmod is good, 664+775 or 660+770
    ### Doing directories ... when to stop ? 2 dirs ?
    ### the makedirs is supposed to take a 'mode' argument, that didn't work for me.
    chmod664 = stat.S_IRUSR | stat.S_IWUSR | stat.S_IRGRP | stat.S_IWGRP | stat.S_IROTH
    chmod775 = stat.S_IRUSR | stat.S_IWUSR | stat.S_IXUSR | \
               stat.S_IRGRP | stat.S_IWGRP | stat.S_IXGRP | \
               stat.S_IROTH | stat.S_IXOTH
    chmod660 = stat.S_IRUSR | stat.S_IWUSR | stat.S_IRGRP | stat.S_IWGRP
    chmod770 = stat.S_IRUSR | stat.S_IWUSR | stat.S_IXUSR | \
               stat.S_IRGRP | stat.S_IWGRP | stat.S_IXGRP
    chmod664 = None

    cmd = [scr_halt_cntl, '-f', halt_file]
    cmd.extend(halt_conditions)
    output, rc = runproc(cmd, getstdout=True, getstderr=True, verbose=verbose)
    if rc != 0:
      if output is not None:
        print(output[1].strip())
      print('scr_halt: ERROR: Failed to update halt file for ' + d)
      ret = 1
    # TODO For file permissions ?
    # Doing 664+775
    #else: # rc == 0:
    elif chmod664 is not None:
      # the file
      try:
        # 664 OR with whatever was there before
        os.chmod(halt_file, chmod664 | os.stat(halt_file).st_mode)
      except:
        pass
      chdir = '/'.join(halt_file.split('/')[:-1])
      try:
        # do both directories: d / .scr
        os.chmod(chdir, chmod775 | os.stat(chdir).st_mode)
        chdir = '/'.join(chdir.split('/')[:-1])
        os.chmod(chdir, chmod775 | os.stat(chdir).st_mode)
      except:
        pass

    # print output to screen
    if output is not None:
      print(output[0].strip())

  # kill job if immediate was set
  # TODO: would like to protect against killing a job in the middle of a checkpoint if possible
  if immediate:
    # TODO: lookup active jobid for given prefix directory and halt job based on system
    print('scr_halt: ERROR: --immediate option not yet supported')
    return 1

  return ret


if __name__ == '__main__':
  parser = argparse.ArgumentParser(
      add_help=False,
      argument_default=argparse.SUPPRESS,
      prog='scr_halt',
      epilog=
      'TIME arguments are parsed using parsetime.py,\nand t may be specified in one of many formats.\nExamples include \'12pm\', \'yesterday noon\', \'12/25 15:30:33\', and so on.\nIf no directory is specified, the current working directory is used.'
  )
  # when prefixes are unambiguous then also adding shortcodes isn't necessary
  #parser.add_argument('-z', '--all', action='store_true', help='Halt all jobs on the system.')
  #parser.add_argument('-u', '--user', metavar='LIST', type=str, help='Halt all jobs for a comma-separated LIST of users.')
  parser.add_argument('-c',
                      '--checkpoints',
                      metavar='N',
                      default=None,
                      type=int,
                      help='Halt job after N checkpoints.')
  parser.add_argument(
      '-b',
      '--before',
      metavar='TIME',
      default=None,
      type=str,
      help='Halt job before specified TIME. Uses SCR_HALT_SECONDS if set.')
  parser.add_argument('-a',
                      '--after',
                      metavar='TIME',
                      default=None,
                      type=str,
                      help='Halt job after specified TIME.')
  parser.add_argument('-i',
                      '--immediate',
                      action='store_true',
                      default=False,
                      help='Halt job immediately.')
  parser.add_argument('-s',
                      '--seconds',
                      metavar='N',
                      default=None,
                      type=str,
                      help='Set or reset SCR_HALT_SECONDS for active job.')
  parser.add_argument(
      '-l',
      '--list',
      action='store_true',
      default=False,
      help='List the current halt conditions specified for a job or jobs.')
  parser.add_argument('--unset-checkpoints',
                      action='store_true',
                      default=False,
                      help='Unset any checkpoint halt condition.')
  parser.add_argument('--unset-before',
                      action='store_true',
                      default=False,
                      help='Unset any halt before condition.')
  parser.add_argument('--unset-after',
                      action='store_true',
                      default=False,
                      help='Unset halt after condition.')
  parser.add_argument('--unset-seconds',
                      action='store_true',
                      default=False,
                      help='Unset halt seconds.')
  parser.add_argument('--unset-reason',
                      action='store_true',
                      default=False,
                      help='Unset the current halt reason.')
  parser.add_argument('-r',
                      '--remove',
                      action='store_true',
                      default=False,
                      help='Remove halt file.')
  parser.add_argument('-v',
                      '--verbose',
                      action='store_true',
                      default=False,
                      help='Increase verbosity.')
  parser.add_argument('-h',
                      '--help',
                      action='store_true',
                      help='Show this help message and exit.')
  parser.add_argument('dirs', nargs=argparse.REMAINDER, default=[])
  args = vars(parser.parse_args())

  if 'help' in args:
    parser.print_help()
  #elif 'all' in args or 'user' in args:
  #  print('scr_halt: ERROR: --all and --user options not yet supported')
  else:
    ret = scr_halt(checkpoints=args['checkpoints'],
                   before=args['before'],
                   after=args['after'],
                   immediate=args['immediate'],
                   seconds=args['seconds'],
                   dolist=args['list'],
                   unset_checkpoints=args['unset_checkpoints'],
                   unset_before=args['unset_before'],
                   unset_after=args['unset_after'],
                   unset_seconds=args['unset_seconds'],
                   unset_reason=args['unset_reason'],
                   remove=args['remove'],
                   verbose=args['verbose'],
                   dirs=args['dirs'])
    print(str(ret))
